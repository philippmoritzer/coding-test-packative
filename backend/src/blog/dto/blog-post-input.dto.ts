import { ApiProperty } from '@nestjs/swagger';
import { IsNotEmpty, IsString, MaxLength, MinLength } from 'class-validator';

export class BlogPostInputDTO {
  @ApiProperty({
    description: 'Title of the blog post',
    minLength: 3,
    maxLength: 255,
  })
  @IsString()
  @IsNotEmpty()
  @MinLength(3)
  @MaxLength(255)
  title: string;

  @ApiProperty({
    description: 'Description of the blog post',
    minLength: 3,
    maxLength: 255,
  })
  @IsString()
  @IsNotEmpty()
  @MinLength(3)
  @MaxLength(255)
  description: string;
}
