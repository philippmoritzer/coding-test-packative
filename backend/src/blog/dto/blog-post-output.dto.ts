import { ApiProperty } from '@nestjs/swagger';

export class BlogPostOutputDTO {
  @ApiProperty({ description: 'Unique identifier for the blog post' })
  id: string;

  @ApiProperty({ description: 'Title of the blog post' })
  title: string;
  @ApiProperty({ description: 'Content of the blog post' })
  description: string;
  @ApiProperty({ description: 'Number of likes for the blog post' })
  likes: number;
}

export class BlogPostsListOutputDTO {
  @ApiProperty({ type: [BlogPostOutputDTO] })
  posts: BlogPostOutputDTO[];
}
