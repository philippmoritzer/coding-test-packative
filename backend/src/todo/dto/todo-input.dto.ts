/* eslint-disable @typescript-eslint/no-unsafe-call */
import {
  IsEnum,
  IsNotEmpty,
  IsString,
  MaxLength,
  MinLength,
} from 'class-validator';
import { ETodos } from '../interfaces/todo.interface';
import { ApiProperty } from '@nestjs/swagger';

export class TodoItemInputDto {
  @ApiProperty({
    description: 'Name of the todo item',
    example: 'Buy groceries',
    minLength: 3,
    maxLength: 255,
  })
  @IsString()
  @IsNotEmpty()
  @MinLength(3)
  @MaxLength(255)
  name: string;

  @ApiProperty({
    description: 'Type of the todo item',
    enum: ETodos,
    example: ETodos.EMPTY_DISHWASHER,
  })
  @IsEnum(() => ETodos)
  type: ETodos;
}

export class TodoInputDto {
  @IsString()
  @IsNotEmpty()
  @MinLength(3)
  @MaxLength(255)
  name: string;
}
