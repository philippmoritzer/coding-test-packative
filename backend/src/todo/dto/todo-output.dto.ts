import { ApiProperty } from '@nestjs/swagger';
import { ETodos } from '../interfaces/todo.interface';

export class TodoOutputDtoItem {
  @ApiProperty({
    description: 'Unique identifier for the todo item',
    example: '12345',
    minLength: 3,
    maxLength: 255,
  })
  id: string;
  @ApiProperty({
    description: 'Type of the todo item',
    enum: ETodos,
    example: 'EMPTY_DISHWASHER',
  })
  type: string;

  @ApiProperty({
    description: 'Name of the todo item',
    example: 'Buy groceries',
    minLength: 3,
    maxLength: 255,
  })
  name: string;
}

export class TodoOutputDtoList {
  @ApiProperty({
    description: 'List of todo items',
    type: [TodoOutputDtoItem],
  })
  todoItems: TodoOutputDtoItem[];
}
