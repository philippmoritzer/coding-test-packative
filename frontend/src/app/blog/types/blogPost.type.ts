export type BlogPost = {
  id: string;
  title: string;
  content: string;
  createdBy: string;
  likes: number;
  createdAt: string;
  updatedAt: string;
};

export type BlogPostCardProps = {
  blogPost: BlogPost;
};
