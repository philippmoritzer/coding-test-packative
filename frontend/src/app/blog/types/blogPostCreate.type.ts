export type BlogPostCreateType = {
  title: string;
  content: string;
  createdBy: string;
};
