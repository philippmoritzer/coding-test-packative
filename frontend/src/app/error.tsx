'use client';

import React from "react";
import Link from "next/link";

export default function Error({ error, reset }: { error: Error; reset: () => void }) {
  return (
    <div className="min-h-screen flex flex-col items-center justify-center bg-gradient-to-r from-gray-100 to-gray-200">
      <div className="bg-white shadow-lg rounded-lg p-10 max-w-lg w-full text-center">
        <h1 className="text-4xl font-bold text-red-600 mb-4">Something went wrong</h1>
        <p className="text-gray-700 mb-6">
          {error?.message || "An unexpected error has occurred. Please try again later."}
        </p>
        <div className="flex flex-col sm:flex-row gap-4 justify-center">
          <button
            onClick={reset}
            className="bg-blue-600 hover:bg-blue-700 text-white px-6 py-2 rounded transition"
          >
            Try Again
          </button>
          <Link
            href="/blog"
            className="bg-gray-200 hover:bg-gray-300 text-gray-800 px-6 py-2 rounded transition"
          >
            Back to Blog
          </Link>
        </div>
      </div>
    </div>
  );
}