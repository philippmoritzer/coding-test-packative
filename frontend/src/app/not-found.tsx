'use client';

import React from "react";
import Link from "next/link";

export default function NotFound() {
  return (
    <div className="min-h-screen flex flex-col items-center justify-center bg-gradient-to-r from-gray-100 to-gray-200">
      <div className="bg-white shadow-lg rounded-lg p-10 max-w-lg w-full text-center">
        <h1 className="text-5xl font-bold text-blue-600 mb-4">404</h1>
        <h2 className="text-2xl font-semibold text-gray-800 mb-2">Page Not Found</h2>
        <p className="text-gray-700 mb-6">
          Sorry, the page you are looking for does not exist or has been moved.
        </p>
        <Link
          href="/blog"
          className="bg-blue-600 hover:bg-blue-700 text-white px-6 py-2 rounded transition"
        >
          Back to Blog
        </Link>
      </div>
    </div>
  );
}