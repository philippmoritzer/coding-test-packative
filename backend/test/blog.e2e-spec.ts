/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { App } from 'supertest/types';
import { AppModule } from '../src/app.module';
import { redactDynamicFields } from './test.helper';
import { BlogPostOutputDTO } from 'src/blog/dto/blog-post-output.dto';
import { randomUUID } from 'crypto';

describe('AppController (e2e)', () => {
  let app: INestApplication<App>;
  let mockPost: BlogPostOutputDTO;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();

    //Create a mock post
    mockPost = (
      await request(app.getHttpServer())
        .post('/blog/posts')
        .set('Authorization', 'userId:some-userid;permissions:create:blog-post')
        .send({
          title: 'Test Post',
          description: 'This is a test post',
        })
        .expect(201)
    ).body;
  });

  afterAll(async () => {
    await app.close();
  });

  it('/blog/posts (GET) Blog posts', async () => {
    const res = await request(app.getHttpServer())
      .get('/blog/posts')
      .set('Authorization', 'userId:some-userid')
      .expect(200)
      .expect((res) => {
        console.log(res.body);
        expect(res.body.posts).toBeInstanceOf(Array);
        expect(res.body.posts.length).toBeGreaterThanOrEqual(0);
      });

    expect(redactDynamicFields(res.body)).toMatchSnapshot();
  });

  it('/blog/posts (GET) 401 on get posts', async () => {
    await request(app.getHttpServer())
      .get('/blog/posts')
      .expect(401)
      .expect((res) => {
        expect(res.body).toHaveProperty('statusCode', 401);
        expect(res.body).toHaveProperty(
          'message',
          'Authorization header is missing',
        );
      });
  });

  it('/blog/posts (POST) Create a blog post', async () => {
    const res = await request(app.getHttpServer())
      .post('/blog/posts')
      .set('Authorization', 'userId:some-userid;permissions:create:blog-post')
      .send({
        title: 'Test Post',
        description: 'This is a test post',
      })
      .expect(201)
      .expect((res) => {
        expect(res.body).toHaveProperty('id');
        expect(res.body.title).toBe('Test Post');
        expect(res.body.description).toBe('This is a test post');
      });

    expect(redactDynamicFields(res.body)).toMatchSnapshot();
  });

  it('/blog/posts (POST) 403 on post', async () => {
    await request(app.getHttpServer())
      .post('/blog/posts')
      .set('Authorization', 'userId:some-userid;permissions:read:blog-post')
      .send({
        title: 'Test Post',
        description: 'This is a test post',
      })
      .expect(403)
      .expect((res) => {
        expect(res.body).toHaveProperty('statusCode', 403);
        expect(res.body).toHaveProperty(
          'message',
          'User does not have the required permissions',
        );
      });
  });

  it('/blog/posts/:id/like (PATCH) Like a blog post', async () => {
    const res = await request(app.getHttpServer())
      .patch(`/blog/posts/${mockPost.id}/like`)
      .set('Authorization', 'userId:some-userid;')

      .expect(200)
      .expect((res) => {
        expect(res.body).toHaveProperty('likes');
        expect(res.body.likes).toBe(1);
      });
    expect(redactDynamicFields(res.body)).toMatchSnapshot();
  });
  it('/blog/posts/:id/like (PATCH) validate param blog post id', async () => {
    await request(app.getHttpServer())
      .patch('/blog/posts/invalid-id/like')
      .set('Authorization', 'userId:some-userid;')
      .expect(400)
      .expect((res) => {
        expect(res.body).toHaveProperty('statusCode', 400);
      });
  });
  it('/blog/posts/:id/like (PATCH) 404 blogpost not found', async () => {
    await request(app.getHttpServer())
      .patch(`/blog/posts/${randomUUID().toString()}/like`)
      .set('Authorization', 'userId:some-userid;')
      .expect(404)
      .expect((res) => {
        expect(res.body).toHaveProperty('statusCode', 404);
        expect(res.body).toHaveProperty('message', 'Post not found');
      });
  });
});
