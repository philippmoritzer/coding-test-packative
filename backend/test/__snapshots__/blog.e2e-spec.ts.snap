// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) /blog/posts (GET) Blog posts 1`] = `
{
  "posts": [
    {
      "createdAt": "[REDACTED]",
      "createdBy": "some-userid",
      "description": "This is a test post",
      "id": "[REDACTED]",
      "likes": 0,
      "title": "Test Post",
      "updatedAt": "[REDACTED]",
    },
  ],
}
`;

exports[`AppController (e2e) /blog/posts (POST) Create a blog post 1`] = `
{
  "createdAt": "[REDACTED]",
  "createdBy": "some-userid",
  "description": "This is a test post",
  "id": "[REDACTED]",
  "likes": 0,
  "title": "Test Post",
  "updatedAt": "[REDACTED]",
}
`;

exports[`AppController (e2e) /blog/posts/:id/like (PATCH) Like a blog post 1`] = `
{
  "createdAt": "[REDACTED]",
  "createdBy": "some-userid",
  "description": "This is a test post",
  "id": "[REDACTED]",
  "likes": 1,
  "title": "Test Post",
  "updatedAt": "[REDACTED]",
}
`;
